src/gui/raster/qgscolorrampshaderwidget.cpp

```diff
@@ -356,6 +356,44 @@ void QgsColorRampShaderWidget::mClassificationModeComboBox_currentIndexChanged(
   emit classificationModeChanged( mode );
 }
 
+void QgsColorRampShaderWidget::updateColorRamp()
```

Comment:
How would you feel about moving this logic to QgsColorRampShader ? We could potentially:

1.Mmove this to QgsColorRampShader::createRamp()
2. Call createRamp here, so that we still get the nice updating of the source color ramp widget
3. Use QgsColorRampShader::createRamp instead of sourceColorRamp wherever we are creating QgsColorRampLegendNode, so that the legend node uses the updated ramp

The advantage would be that programatic updates to the shader colors/values would also result in the legend correctly updating, instead of just GUI interactions.